### ISSUE TITLE
Segfault when creating from array

### ISSUE THREAD

**Issue by dkam** [libvips/ruby-vips#249]
I assume it should be possible to create an image like this:

```image = Vips::Image.new_from_array( Vips::Image.new_from_file('600.jpeg').to_a )```

```shell
/Users/dkam/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/ruby-vips-2.0.17/lib/vips/image.rb:675: [BUG] Segmentation fault at 0x0000000000000128
ruby 2.7.1p83 (2020-03-31 revision a0c7c23c9c) [x86_64-darwin19]

-- Crash Report log information --------------------------------------------
   See Crash Report log file under the one of following:                    
     * ~/Library/Logs/DiagnosticReports                                     
     * /Library/Logs/DiagnosticReports                                      
   for more details.                                                        
Don't forget to include the above Crash Report log file in bug reports.     
```

This is on MacOSX 10.15.6.  It also crashes on Ubuntu 18.04.5 LTS.  Happen with multiple Ruby versions.

**Comment by jcupitt** [comment]
Hello @dkam,

`new_from_array` only works for monochrome images. It's supposed to be for making small constant images, like structuring elements for the morphological operators. For example:

```ruby
mask = Vips::Image.new_from_array([
  [128, 255, 128],
  [255, 255, 255],
  [128, 255, 128],
])
```

`#to_a` works for RGB images, but it'll make a 3D array, which seems to be what's causing the crash. I'll add a check, thanks for the report!

Why do you need to work with images as huge arrays? Perhaps there's another (more efficient) way to solve your problem.

**Comment by dkam** [comment]
Thanks!  I'm trying to embed a book cover with a drop shadow into a background image.  I thought I'd build up the background image by creating a massive array.  It'd be nice to be able to get a gradient background too. 

![image](https://user-images.githubusercontent.com/18232/93564638-5e10e980-f9cd-11ea-8fd0-4da4093eb9d5.png)

I think I can build the image I want with SVG but using ```vips copy image.svg image.jpg``` results in an empty image. I suspect it's because I was including the book cover via  Xlink:href and rsvg didn't like it. 

I'm having trouble figuring out how Vips::Image#ifthenself works - but this is what I have so far: 

```ruby
back  = Vips::Image.new_from_file 'background.jpg'
cover = Vips::Image.new_from_file('600.jpeg')
scover = cover.resize( 300/cover.width.to_f )

emb = scover.embed ((back.width - scover.width)/2), 100, back.width, back.height

out = back.ifthenelse [255, 0, 0], emb, :blend => false
```

If I manage to get that working, I think my next task would be to create a rectangle the size of the cover, apply a Gaussian blur and layering it between the background and the cover.

Any tips you have would be appreciated! 

Thanks,

**Comment by dkam** [comment]
Hey @jcupitt I figured out how to make an image of a single colour and documented it in the [wiki](https://github.com/libvips/ruby-vips/wiki/Examples).  Hope that's ok!  Any help on the other stuff would be great.

**Comment by jcupitt** [comment]
That's right, you can make a background, a shadow and the image, then composite them together.

For example:

```ruby
#!/usr/bin/ruby

require "vips"

image = Vips::Image.new_from_file ARGV[0], access: :sequential

background_width = 1200
background_height = 700
shadow_radius = 40
image_left = (background_width - image.width) / 2
image_top = (background_height - image.height) / 2
shadow_colour = [128]
background_colour = [241, 237, 247]

# make the shadow mask ... we have to enlarge before we blur or the soft edges
# will be clipped
# gaussblur take sigma as an argument ... approximate as radius / 2
shadow_mask = image.
  new_from_image(255).
  copy(interpretation: :b_w).
  embed(shadow_radius, shadow_radius, 
        image.width + 2 * shadow_radius,
        image.height + 2 * shadow_radius).
  gaussblur(shadow_radius / 2)

# make the background: a block of solid colour
background = image.
  new_from_image(background_colour).
  embed(0, 0, background_width, background_height, extend: :copy)

# make the shadow: a block of solid colour, but with the shadow mask as the
# alpha
shadow = shadow_mask.
  new_from_image(shadow_colour).
  bandjoin(shadow_mask)

# composite the three layers together
image = background.composite([shadow, image], "over", 
                        x: [image_left - shadow_radius, image_left], 
                        y: [image_top - shadow_radius, image_top]) 

image.write_to_file ARGV[1]
```

Run like this:

```
$ ./drop_shadow.rb ~/pics/dune.png x.png
```

To make:

![x](https://user-images.githubusercontent.com/580843/93590023-b3d19b80-f9a5-11ea-8f5f-0f1248b63c9a.png)

**Comment by jcupitt** [comment]
Have you found nip2? It's handy for debugging scripts since it can display all the vips image types and their metadata.

You can write something like:

```ruby
def view(image)
    image.write_to_file("/tmp/x.v")
    system "nip2 /tmp/x.v ; rm /tmp/x.v")
end
```

And then insert that in your script to check intermediate values.

**Comment by jcupitt** [comment]
I added the extra arg check to `new_from_array` -- it'll be in 2.0.18. Thanks for reporting this!

**Comment by dkam** [comment]
Thanks very much for the example - I've got that all working very nicely. I've added a simple version to the examples page on the wiki.   I'm adding a rotate now with this before the compose:

```ruby
shadow = shadow.rotate(15)
image = image.bandjoin(255).rotate(15)
```

Without the bandjoin, the empty space is black, so the bandjoin(255) is adding a alpha channel?  The doc says it's doing a band-wise join - but I don't see how 255 is an array of images - how does that work?

**Comment by jcupitt** [comment]
You can use a constant anywhere you can use an image and it (effectively) becomes an image of the correct size.

`bandjoin(255)` means add an extra band with 255 in every pixel. You can write `.bandjoin([255, 255])` to add two extra bands. You can write things like `(image < 128).ifthenelse(255, 128)` etc.

**Comment by Nakilon** [comment]
I wish `new_from_array` was not segfaulting. In my case I build the colourfull image line by line. What are alternatives? Ruby can't do multidimensional transposing like Mathematica. Seems like I have to return three lines per line and then transpose and `bandjoin`.

**Comment by jcupitt** [comment]
The segv should be fixed in git master ruby-vips.

There's a PR to add `new_from_memory`, you probably saw:

https://github.com/libvips/ruby-vips/pull/253

That would probably be the best way to pass large binary images into ruby-vips.

**Comment by Nakilon** [comment]
> PR to add `new_from_memory`

Oh, if I understood correctly, that looks like if ruby-vips user has some raw binary data blobs. Like ASCII-encoded string with channel values encoded in chars.

But in my case it would need some additional `.pack`/`.unpack` routines because I build lines of pixels as pure Ruby array of pixels. Specifically it was a glitch-art:

![image](https://user-images.githubusercontent.com/2870363/95048904-5baecf00-06f1-11eb-94d8-e726a28e7b90.png)

I called `Vips::Image#to_a .map{ |line| ... }`, so I could do things similar to `line[10..50] = line[10..50].reverse` that is easy in pure Ruby and I was able to do it based on calculations on pixel values taken from the same pixel line array.
